kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PaginatedScanSuite.scala

```diff
@@ -85,59 +85,149 @@ class PaginatedScanSuite extends AnyFunSuite with TestUtilsWithTableManagerAPIs
-    validatePageToken(nextPageToken, testCase.expLogFile, testCase.expRowIdx, testName)
+    assert(returnedBatchesInPage.nonEmpty)
+
+    val fileCounts: Seq[Long] = returnedBatchesInPage.map(_.getPreComputedNumSelectedRows
+      .get().toLong)
+    val totalFileCountsReturned = fileCounts.sum
+
+    logger.info(s"number of batches = ${returnedBatchesInPage.length}")
+    logger.info(s"number of AddFiles = ${totalFileCountsReturned}")
+
+    if (nextPageToken.isPresent) {
+      val lastReadLogFilePath = PageToken.fromRow(nextPageToken.get).getLastReadLogFilePath
+      val lastReturnedRowIndex = PageToken.fromRow(nextPageToken.get).getLastReturnedRowIndex
+
+      logger.info(s"New PageToken: lastReadLogFileName = $lastReadLogFilePath")
+      logger.info(s"New PageToken: lastReadRowIndex = $lastReturnedRowIndex")
+    }
 
-    nextPageToken
+    (nextPageToken, returnedBatchesInPage)
   }
 
-  // TODO: test call hasNext() twice
+  /**
+   * Simulates the client's behavior of reading a full scan in a paginated manner
+   * with a given page size.
+   *
+   *  The client:
+   * 1. Starts by requesting the first page (no page token).
+   * 2. Receives a page of results along with a page token.
+   * 3. Uses the page token to request the next page.
+   * 4. Repeats until the returned page token is empty, indicating that all data has been consumed.
+   */
+  private def runCompletePaginationTest(
+      testCase: SinglePageRequestTestCase,
+      tablePath: String,
+      tableVersionOpt: Optional[Long] = Optional.empty()): Unit = {
+
+    // ============ Request the first page ==============
+    var (pageTokenOpt, returnedBatchesInPage) = doSinglePageRequest(
+      tablePath = tablePath,
+      tableVersionOpt = tableVersionOpt,
+      pageSize = testCase.pageSize)
+
+    validateFirstPageResults(
+      returnedBatchesInPage,
+      testCase.expFirstPageFileCnt,
+      testCase.expFirstPageBatchCnt)
+
+    // When the scan is exhausted, returned page token should be empty.
+    if (pageTokenOpt.isPresent) {
+      validateFirstPageToken(
+        pageTokenOpt.get,
+        testCase.expLastReadLogFileInFirstPage,
+        testCase.expLastReadRowIdxInFirstPage)
+    }
+
+    // ============ Request following pages ==============
+    var allBatchesPaginationScan = returnedBatchesInPage
+    while (pageTokenOpt.isPresent) {
+      val (newPageTokenOpt, newReturnedBatchesInPage) = doSinglePageRequest(
+        tablePath = tablePath,
+        tableVersionOpt = tableVersionOpt,
+        pageTokenOpt = pageTokenOpt,
+        pageSize = testCase.pageSize)
+      pageTokenOpt = newPageTokenOpt
+      allBatchesPaginationScan ++= newReturnedBatchesInPage
+    }
+
+    val normalScan =
+      getScanBuilder(tablePath = tablePath, tableVersionOpt = tableVersionOpt).build()
+
+    val iter = normalScan.getScanFiles(customEngine)
+    val allBatchesNormalScan = collectPaginatedBatches(iter)
```

Comment:
you are calling `collectPaginatedBatches` but the `iter` is from a normal scan? perhaps your helper method should just be called `collectBatches`?

btw, `CloseableIterator` has a method on it called `toinMemoryList` you could probably use that?

https://github.com/delta-io/delta/blob/8b7fd6d67d33c321f1751d44103b90e187285867/kernel/kernel-api/src/main/java/io/delta/kernel/utils/CloseableIterator.java#L260